### PR TITLE
Globalization test updates

### LINKED ISSUE #9386: Enhancements to @bind - globalization and conversions
**Edit: @rynowak hijacking the top post for great justice**

# Summary

`@bind` when used with `<input ...>` elements is the primary way that users parse and convert user input into .NET types.  This makes `@bind` the primary touch point for globalization - dealing with input and display of information in a locale specific way.

Goals:
- Support users using any locale in client-side/server-side Blazor
- Support users using many locales with a single server for server-side Blazor
- Support flexible custom conversions and formats
- Provide good support for common built-in .NET types

**TLDR My recommendations for 3.0:**
- [x] Make sure user's culture is reflected in .NET
   - ~~Client-side~~ will be done post 3.0
   - [x] Prerendering
   - [x] Server-side (circuits)
- [x] Update runtime to accept culture 
- [x] Update compiler to support culture and format in mappings
- [x] Update `type="..."` mappings to support culture and formats
- [x] Make our form controls work correctly with different locales
- [x] Add `@bind` support for:
    - [x] `DateTime?`
    - [x]  `DateTimeOffset`
    - [x]  `DateTimeOffset?`
- [x] Support `TypeConverter`

**CUT**
- Format strings for numbers
- Unusual primitive types like `byte`
- Our own converter primitive (use `TypeConverter`)

## Related user-feedback
 - https://github.com/aspnet/AspNetCore/issues/5471 Runtime Culture and Localization
- https://github.com/aspnet/AspNetCore/issues/5523 Add support for binding decimal, double, float data type with configurable formatting
- https://github.com/aspnet/AspNetCore/issues/5651 'bind' does not accept values of type System.DateTimeOffset 
- https://github.com/aspnet/AspNetCore/issues/7655 Razor Components (Blazor 0.8) - "double/float/decimal" error parse
- https://github.com/aspnet/AspNetCore/issues/8203 Razore Component Doesn't Support "short" Data Type
- https://github.com/aspnet/AspNetCore/issues/8493 TypeConverter for parameters 
- https://github.com/aspnet/AspNetCore/issues/9435 Blazor Custom formatting for numbers and strings 
- https://github.com/aspnet/AspNetCore/issues/9632 typeparam bind error 
- https://github.com/aspnet/AspNetCore/issues/9939 Blazor form component InputSelect should support binding of type Guid 
- https://github.com/aspnet/AspNetCore/issues/10077 EventCallback<T> Error cannot convert from 'method group' to 'EventCallback'
- https://github.com/aspnet/AspNetCore/issues/8385 EventCallback not working as expected

## Challenges

HTML5 field definitions make dealing with globalization more complicated, because depending on the field type, the culture-sensitivity expectations are different.

```razor
<input type="text">Enter a number in German</input>
<input type="number">Enter a number in German</input>
```

In the `type="text"` field the `.value` attribute will contain exactly what the user typed in - this implies that the value needs to parsed, or formatted according the users' locale.

In the `type="number"` field the `.value` attribute will contain a number value in an standard non-locale-specific format. The value must be parsed and formatted according to this standard format. It's up to the browser to support the users' preferred style of display for their locale.

It stands to reason that users will want to use `text` fields for numbers, dates, etc in some cases instead of the HTML5 equivalents. Using `text` is much more flexible since it's essentially free-form text, as opposed to `number` which is restricted by the browser, and has a set of UI behaviors the developer might not want.

Additionally, for the server-side Blazor programming model, there's no automatic way to flow the culture from the client to the server. We have to build it in to Blazor. 

## Plan of record

Since some fields require the use of `CultureInfo.InvariantCulture` and some require hardcoded formats (`type="month"`) we need to make our parsing and formatting code aware of these details. This knowledge needs to either come from the runtime (JS or .NET), the compiler, or the user.

At first glance it seems like it would be straightforward to put this knowledge in the JS part of the runtime. At this layer we can see each DOM element and it's attributes, so we should be able to tell what type of field we're interacting how. However, this has a few drawbacks. First, we'd need to pull in some JS glob/loc packages, which would increase the size. Secondly, we'd be hoping and assuming that these JS libraries have the same behaviors as .NET. That doesn't seem reasonable.

So we settled on a hybrid approach. We'll allow the user to specify the culture or format, but try to get it right for the known `type="..."` values. This means that if we see `type="number"` statically in code, we can apply the right culture. If the user wants to assign the field type dynamically then they will need to specify the format or culture manually.

### More control over `@bind`

Some ideas here... we don't have to do all of these, and probably shouldn't try to do all of these in the first release.

---

**Flow the field type into .NET:** We could easily capture data about the field type in the JS side of the binding code. This would allow the .NET code to be aware of what the `type="..."` attribute specifies - which we could use to perform the conversion using invariant culture or the user's culture. With this done, we could make the default be culture-sensitive, and apply invariant culture where appropriate based on the field type.

---

**Allow users to specify the culture for `@bind`** We could allow users to specify the culture at the call-site. This could be a desirable level of control for application developers, but it's not a good default experience - essentially we'd be picking a default behavior and allowing an override. This isn't a good default user experience by itself -  because as we've already seen, HTML5 fields like `number` want a different default than `text.

Directive attributes make it easy for us to build something like this

```razor
<input type="text" @bind="coolnessLevel" @bind:culture="CultureInfo.InvariantCulture">
  Rate the coolness of Blazor (in a non-locale-specific way)
</input>
```

---

**Allow format strings for more types** Currently we support format strings for `DateTime` and only datetime. This is useful because `DateTime` is used to represent a variety of different *kinds of values*, times, dates, both, etc.  This is also useful because `DateTime` supports a format string on **both** formatting and parsing.

It might seem like allowing format strings on more types will free us from the need to do some of these other solutions - however:
- Format strings in .NET are non-locale-specific.
- If you could write a format string is locale-specific, then that ties you to a single locale.
- Types other than `DateTime` don't use format strings for parsing. (ouch)

So while format strings might be useful, they don't really solve any globalization related problems. 

Its definitely possible to support format strings for more types, but we need to make sure users understand that the format string they specify is only used in formatting, and not in parsing. The parsing behaviour is specified by [`NumberStyles`](https://docs.microsoft.com/en-us/dotnet/api/system.int32.parse?view=netframework-4.8#System_Int32_Parse_System_String_System_Globalization_NumberStyles_System_IFormatProvider_). This implies that we need to be really permissive in what we parse if we want users to specify a format for numbers.

---

**Add support for missing types** We could add support to `bind` for types that don't currently support like `DateTimeOffset`, `DateTime?` and a bunch of the rarely-used integral types. Look at the types supported by `System.Convert`

Since `bind` currently only supports enums and the types we hardcode we don't have support for types like `ushort` or `byte`. It seems obvious to me that there's a priority order of these things:
- `DateTimeOffset` & `DateTimeOffset?` & `DateTime?`
- everything else 

I think an important question when you consider more exotic types like `char` or `sbyte` is - *what is the expectation of someone who reaches for that type?*. I don't think we know. It's easy to try and understand the motivation of someone using `DateTime` or `int` or `decimal`. 

It gives me a little bit of pause when I think about this because I don't know what that user wants - do they want Hex? Octal? Why?

Paired with adding support for type-converters, we could end up in a scenario where not providing hand-coded `bind` support in this release and adding it in a future release results in a breaking change.

---

**Add support for type converters** We could easily add support for arbitrary types to *fall back* to type converters. Type convert is a well-known extensibility point that .NET provides, and other parts of ASP.NET Core use for conversions.

This seems like a clear win, because we get support for things like `Guid` and `TimeSpan` without writing much additional code. Users can write their own `TypeConverter` and plug it in using a non-Blazor mechanism and it will *just work*.

This is actually somewhat high priority because users hit this trying to use `@bind` with a generic type. Since we don't overload `@bind` for arbitrary types, this will result in a compile error and a bad experience.

---

**Allow users to specify a custom converter`** We could allow users to specify a converter at the call-site. This could be a desirable level of control for application developers when they are doing something complex.

Directive attributes make it easy for us to build something like this

```razor
<input type="text" @bind="coolnessLevel" @bind:converter="MyConverter.Default">
  Rate the coolness of Blazor (in a totally custom way)
</input>
```

However, this doesn't really solve any globalization problems. This is just added flexibility.

### Flow the culture 

We have to make sure that .NET code is running with the correct `CultureInfo.CurrentCulture` and `CultureInfo.CurrentUICulture`. This is required for localization (using `.resx` as well as doing any kind of culture-specific conversion).

Having the properties reflect the user's locale is essential for making Blazor *feel like .NET* because without this, a bunch of .NET infrastructure will work incorrectly. You normally accomplish this in .NET using the localization middleware and `Accept-Language`, but I'm not sure that applies to our scenario. *Asterix 1*

Concerns that need to be handled:
- How does this work in client-side?? We want this to be *free* and provided by Mono
- How does this work in prerendering?? (localization middleware)
- How does this work with Circuits?? (I assume we have to build this)

This of course is considering that we can assume `navigator.language` reflects what the user wants to see. I think we need to consider providing an override or some other mechanism to customize this to be out of scope for 3.0 and react to feedback.

### Asterix 1

Users **really** like to get fancy with the localization middleware and do all kinds of things that aren't based on `Accept-Language`. Typically means using the query string or URL path. These mechanisms aren't inherently visible to JS in the same way that `Accept-Language` would be - since the browser generates `Accept-Language`. 

I expect we're going to a substantial amount of feedback from users that want something *fancy* to work, but I don't see a way we can make that possible for either Blazor scenario. The trap here is that the localization middleware will be required for pre-rendering, but won't be used for non-prerendering - resulting in subtle bugs in users' applications. The solution would be for users to write their own logic to resolve the culture (in JS).

### LINKED ISSUE #12286: React to test failures from bind + globalization changes

### PR DESCRIPTION
**Pull request by rynowak**
Globalization test updates

Fixes: #9386
Fixes: #12286

### CODE REVIEW

File: src/Components/Blazor/Build/test/BindRazorIntegrationTest.cs [hunk 1]
```diff
@@ -89,7 +89,7 @@ Task IComponent.SetParametersAsync(ParameterCollection parameters)
                 frames,
                 frame => AssertFrame.Component(frame, "Test.MyComponent", 3, 0),
                 frame => AssertFrame.Attribute(frame, "Value", 42, 1),
-                frame => AssertFrame.Attribute(frame, "ValueChanged", typeof(EventCallback<UIChangeEventArgs>), 2));
+                frame => AssertFrame.Attribute(frame, "ValueChanged", typeof(EventCallback<int>), 2));
```
Comment (rynowak) [inline src/Components/Blazor/Build/test/BindRazorIntegrationTest.cs]: The original assertion was wrong here. We want this to generate strong typed stuff that won't try to do conversions. This is an update based on changes to the compiler that fixed the underlying issue.

File: src/Components/Blazor/Build/test/BindRazorIntegrationTest.cs [hunk 2]
```diff
@@ -163,7 +163,7 @@ Task IComponent.SetParametersAsync(ParameterCollection parameters)
                 frames,
                 frame => AssertFrame.Component(frame, "Test.MyComponent", 3, 0),
                 frame => AssertFrame.Attribute(frame, "Value", 42, 1),
-                frame => AssertFrame.Attribute(frame, "OnChanged", typeof(EventCallback<UIChangeEventArgs>), 2));
+                frame => AssertFrame.Attribute(frame, "OnChanged", typeof(EventCallback<int>), 2));
```
Comment (rynowak) [inline src/Components/Blazor/Build/test/BindRazorIntegrationTest.cs]: The original assertion was wrong here. We want this to generate strong typed stuff that won't try to do conversions. This is an update based on changes to the compiler that fixed the underlying issue.

File: src/Components/test/E2ETest/Tests/BindTest.cs [hunk 3]
```diff
@@ -829,10 +829,10 @@ public void CanBindTextboxDateTimeOffsetWithFormat()
             Assert.Equal(expected, DateTimeOffset.Parse(boundValue.Text));
             Assert.Equal(expected, DateTimeOffset.Parse(mirrorValue.GetAttribute("value")));
 
-            // Clear textbox; value updates to emtpy because that's what we do for `default` when there's a format
+            // Clear textbox; value updates to the default
             target.Clear();
             expected = default;
-            Browser.Equal(string.Empty, () => target.GetAttribute("value"));
+            Browser.Equal("01-01", () => target.GetAttribute("value"));
```
Comment (rynowak) [inline src/Components/test/E2ETest/Tests/BindTest.cs]: @SteveSandersonMS - let's discuss if you have concerns

This behaviour was inconsistent between the case where you have a format and where you don't. I made these behave the same, which broke the test. So for `default(DateTime)` we will tostring the value regardless of whether you specified a format. For `default(DateTime?)` we will return the empty string.

File: src/Components/test/E2ETest/Tests/BindTest.cs [hunk 4]
```diff
@@ -770,11 +770,11 @@ public void CanBindTextboxDateTimeWithFormat()
             Assert.Equal(expected, DateTime.Parse(boundValue.Text));
             Assert.Equal(expected, DateTime.Parse(mirrorValue.GetAttribute("value")));
 
-            // Clear textbox; value updates to emtpy because that's what we do for `default` when there's a format
+            // Clear textbox; value updates to the default
             target.Clear();
             target.SendKeys("\t");
             expected = default;
-            Browser.Equal(string.Empty, () => target.GetAttribute("value"));
+            Browser.Equal("01-01", () => target.GetAttribute("value"));
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/Tests/BindTest.cs]: This is a cool behavior. I like it.

File: src/Components/test/E2ETest/Tests/BindTest.cs (hunk 3 above)
Comment (SteveSandersonMS) [reply src/Components/test/E2ETest/Tests/BindTest.cs]: I think that's perfect and matches exactly with the way we started treat numbers (i.e., clearing an `int`-bound input will revert to `0`).

File: src/Components/test/testassets/BasicTestApp/GlobalizationBindCases.razor [hunk 5]
```diff
@@ -30,11 +30,11 @@
 <div>
     <p>Numbers using bind in number fields</p>
     <div>
-        int: <input type="number" id="input_type_number_int" value="@inputTypeNumberInt.ToString(CultureInfo.InvariantCulture)" @onchange="@EventCallback.Factory.CreateBinder(this, (value) => inputTypeNumberInt = value, inputTypeNumberInt, CultureInfo.InvariantCulture)" />
+        int: <input type="number" id="input_type_number_int" @bind="inputTypeNumberInt" />
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/BasicTestApp/GlobalizationBindCases.razor]: Yeah, that's quite an improvement 😄

File: src/Components/test/testassets/TestServer/Startup.cs [hunk 6]
```diff
@@ -76,7 +77,12 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
                     options.AddSupportedCultures("en-US", "fr-FR");
                     options.AddSupportedUICultures("en-US", "fr-FR");
 
-                    // Cookie culture provider is included by default.
+                    // Cookie culture provider is included by default, but we want it to be the only one.
+                    options.RequestCultureProviders.Clear();
+                    options.RequestCultureProviders.Add(new CookieRequestCultureProvider());
+
+                    // We want the default to be en-US so that the tests for bind can work consistently.
+                    options.SetDefaultCulture("en-US");
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/TestServer/Startup.cs]: Does the fact that we use the cookie culture provider mean that, if you've set a non-en-US culture, then the tests will be broken on your machine until you clear that cookie?

I don't have a great alternative suggestion. Maybe the tests that are meant to *only* work in en-US could take care of forcing the thread to en-US in setup and reverting to the previous culture in teardown.

Maybe it's not really a problem. If this fixes things in default cases I'm fine with proceeding as-is, at least until any actual issue becomes evident.

**Review by SteveSandersonMS — APPROVED** [review]
New behavior looks great!

File: src/Components/test/testassets/TestServer/Startup.cs (hunk 6 above)
Comment (rynowak) [reply src/Components/test/testassets/TestServer/Startup.cs]: I don't think the cookies in selenium are durable across tests/runs. We already have tests that flip this setting back and forth between `en-US` and `fr-FR` and it hasn't caused any problems.

**Pull request merged**